common/src/main/java/com/microsoft/identity/common/internal/result/MsalBrokerResultAdapter.java

```diff
@@ -147,15 +187,115 @@ public Bundle bundleFromAuthenticationResult(@NonNull final ILocalAuthentication
+                    .refreshToken(authenticationResult.getRefreshToken())
+                    .refreshTokenAge(authenticationResult.getRefreshTokenAge());
+        }
 
-        return resultBundle;
+        return brokerResultBuilder.build();
+    }
+
+    /**
+     * The Broker-MSAL protocol version which starts supporting AAD RT removal.
+     **/
+    public static final String REMOVE_RT_FROM_AAD_RESULT_MSAL_PROTOCOL_VERSION = "16.0";
+
+    /**
+     * Remove Refresh token from the given {@link ICacheRecord} list.
+     **/
+    @Nullable
+    private List<ICacheRecord> removeRefreshTokenFromCacheRecords(@Nullable final List<ICacheRecord> originalList){
+        if (originalList == null){
+            return null;
+        }
+
+        final List<ICacheRecord> cacheRecordsWithoutRT = new ArrayList<>();
+        for (final ICacheRecord record : originalList) {
+            cacheRecordsWithoutRT.add(
+                    CacheRecord.builder()
+                            .account(record.getAccount())
+                            .idToken(record.getIdToken())
+                            .v1IdToken(record.getV1IdToken())
+                            .accessToken(record.getAccessToken())
+                            .build()
+            );
+        }
+
+        return cacheRecordsWithoutRT;
+    }
+
+    /**
+     * Determine if AAD RT should be removed from the response object.
+     **/
+    private boolean shouldRemoveRefreshTokenFromResult(@NonNull final ILocalAuthenticationResult result,
+                                                      @Nullable final String negotiatedBrokerProtocolVersion){
+        final String methodTag = TAG + ":shouldRemoveRefreshTokenFromResult";
+        if (mShouldStopReturningRtWithAadResponseCallback.getResult()) {
+            // We'll only strip RT if both the client side and broker side supports protocol version 16.
+            if (!isFirstVersionOlderOrEqual(REMOVE_RT_FROM_AAD_RESULT_MSAL_PROTOCOL_VERSION, negotiatedBrokerProtocolVersion)) {
+                SpanExtension.current().setAttribute(
+                        AttributeName.stop_returning_rt_result.name(),
+                        "protocol_version_too_low");
+                return false;
+            }
+
+            try {
+                // Remove RT if the result is not MSA... by checking the "local" TenantID (realm).
+                // (e.g. For the guest account scenario,
+                //       if the account's home tenant is A and it's guested in B...
+                //       this value will be B.
+                if (!StringUtil.isNullOrEmpty(result.getTenantId())){
+                    final boolean resultToReturn = !AzureActiveDirectoryAudience.MSA_MEGA_TENANT_ID.equalsIgnoreCase(result.getTenantId());
+                    SpanExtension.current().setAttribute(
+                            AttributeName.stop_returning_rt_result.name(),
+                            resultToReturn
```

Comment:
What is the data type of `stop_returning_rt_result`? It seems sometimes we are putting boolean but sometimes we are putting String? Let's be consistent and only put 1 type of data in it.

Reply:
will convert all of them to string.